app/src/main/java/net/christianbeier/droidvnc_ng/MainService.java

```diff
@@ -753,12 +766,88 @@ static boolean isServerActive() {
+     * @return An ArrayList&lt;NetworkInterfaces&gt; containing the above-mentioned interfaces
+     */
+    static ArrayList<NetworkInterface> getAvailableNICs() {
+        ArrayList<NetworkInterface> nics = new ArrayList<>();
+
+        try {
+            // Thanks go to https://stackoverflow.com/a/20103869/361413
+            Enumeration<NetworkInterface> nis = NetworkInterface.getNetworkInterfaces();
+            NetworkInterface ni;
+            while (nis.hasMoreElements()) {
+                ni = nis.nextElement();
+                if (ni.isUp()) {
+                    // Check if there are actual ipv4 addresses, if so, add the NetworkInterface
+                    // Should we consider IPv6 also? Technically yes, but the program, at the moment, does not support them.
+                    for (InterfaceAddress ia : ni.getInterfaceAddresses()) {
+                        if (ia.getAddress().getAddress().length == 4) {
+                            nics.add(ni);
+                            break;
+                        }
+                    }
+                }
+            }
+        } catch (SocketException e) {
+            //unused
+        }
+
+        return nics;
+    }
+
+
+
+    /**
+     * Gets all the ipv4s available for a single NetworkInterface
+     * @param ni the NetworkInterface to check
+     * @return an ArrayList&lt;String&gt; which contains all the collected ipv4
+     */
+    static ArrayList<String> getIPv4ForInterface(NetworkInterface ni) {
+        ArrayList<String> ipv4s = new ArrayList<>();
+
+        for (InterfaceAddress ia : ni.getInterfaceAddresses()) {
+            //filter for ipv4/ipv6
+            if (ia.getAddress().getAddress().length == 4) {
+                //4 for ipv4, 16 for ipv6
+                ipv4s.add(ia.getAddress().toString().replaceAll("/", ""));
+            }
+        }
+
+        return ipv4s;
+    }
+
+
+    /**
+     * This returns A SINGLE ipv4 address for the selected interface... giving the possibility
+     * to the server to listen to that address only.
+     * @param ifName name (techName, not friendly one) that defines the interface to be used
+     * @return if the interface 0.0.0.0 or there's some error null, otherwise, the needed ipv4
      */
-    static ArrayList<String> getIPv4s() {
+    static String getInterfaceListeningIPv4Address(String ifName) {
+        if (!ifName.equals("0.0.0.0")) {
+            try {
+                ArrayList<String> ipv4s = MainService.getIPv4ForInterface(NetworkInterface.getByName(ifName));
+                if (ipv4s.size() > 0) {
+                    return ipv4s.get(0);
+                }
+
+            } catch (SocketException ex) {
+                // unused
+            }
+        }
+
+        return null;
+    }
 
+
+    /**
+     * Get all available IPv4 addresses for reaching the current running server.
+     * @return A list of strings, each containing one IPv4 address.
+     */
+    static ArrayList<String> getReachableIPv4s() {
```

Comment:
Can also move to Utils I think

Reply:
I don't think this is a good idea since there's a call to getListenInterface(), which in turns, uses SharedPreferences.
Moreover, there's some specific Chrome OS code.